<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>A single-cell package for Functional Annotation + Gene Module Summarization. • bowerbird</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="A single-cell package for Functional Annotation + Gene Module Summarization.">
<meta property="og:description" content="A single-cell package for Functional Annotation + Gene Module Summarization. Bowerbirds are famous for the elaborate and sometimes whimsical (blue) structures that males build to court females.">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">bowerbird</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/bowerbird_vignette.html">Introduction to bowerbird</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">


<div id="bowerbird" class="section level1">
<div class="page-header"><h1 class="hasAnchor">
<a href="#bowerbird" class="anchor"></a>bowerbird</h1></div>
<p>A single-cell package for Functional Annotation + Gene Module Summarization.</p>
<p><img src="img/logo.png" alt="bower" width="200"></p>
<p><em>Bowerbirds are famous for the elaborate and sometimes whimsical structures that males build to court females.</em></p>
<p><a href="https://blog.nature.org/science/2021/01/04/bowerbirds-meet-the-bird-worlds-kleptomaniac-love-architects/"><img src="https://blog.nature.org/science/files/2020/11/32487196918_8dd537c82a_k.jpg" alt="bower" width="400"></a></p>
<p>Picture credits: A satin bowerbird’s bower, decorated with blue objects. © Stefan Marks / <a href="https://www.flickr.com/photos/stefan_marks/32487196918/">Flickr</a></p>
<p>I’m not sure if this mating ritual is in anyway similar to my attempts to visualise pathway analyses and their association with celltype identities/states… who cares. It’s just a name.</p>
<p>This is being rebuilt from the ground-up in a separate implemenation.</p>
<div id="installation" class="section level2">
<h2 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h2>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># because currently private, requires specifying personal github access token</span>
<span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://devtools.r-lib.org//reference/remote-reexports.html">install_github</a></span><span class="op">(</span><span class="st">'clatworthylab/bowerbird'</span>, auth_token <span class="op">=</span> <span class="st">"insert_your_personal_github_access_token"</span><span class="op">)</span>

<span class="co"># or clone this repo and do devtools::install('path/of/folder/to/bowerbird')</span></code></pre></div>
</div>
<div id="quick-usage" class="section level2">
<h2 class="hasAnchor">
<a href="#quick-usage" class="anchor"></a>Quick Usage</h2>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">bowerbird</span><span class="op">)</span>
<span class="co"># download from msigdb website</span>
<span class="va">gmt_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"h.all.v7.4.symbols.gmt"</span>, package <span class="op">=</span> <span class="st">"bowerbird"</span><span class="op">)</span>
<span class="va">bwr</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/bower.html">bower</a></span><span class="op">(</span><span class="va">gmt_file</span><span class="op">)</span> <span class="co"># this performs a read_geneset step internally, which accepts .gmt, .gmx, .csv, .tsv, .txt, or R objects as list or data.frame format.</span>
<span class="va">bwr</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/snn_graph.html">snn_graph</a></span><span class="op">(</span><span class="va">bwr</span><span class="op">)</span>
<span class="va">bwr</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/find_clusters.html">find_clusters</a></span><span class="op">(</span><span class="va">bwr</span><span class="op">)</span>
<span class="va">bwr</span> <span class="op">&lt;-</span> <span class="fu">summarize_clusters</span><span class="op">(</span><span class="va">bwr</span><span class="op">)</span>
<span class="va">bwr</span>
<span class="co"># BOWER class</span>
<span class="co"># number of genesets:  50</span>
<span class="co"># genesets kNN Graph:</span>
<span class="co"># IGRAPH 5c4d1b0 UNW- 50 124 --</span>
<span class="co"># + attr: name (v/c), cluster (v/n), terms (v/c), labels (v/c), weight (e/n)</span>
<span class="co"># + edges from 5c4d1b0 (vertex names):</span>
<span class="co">#  [1] HALLMARK_TNFA_SIGNALING_VIA_NFKB--HALLMARK_HYPOXIA</span>
<span class="co">#  [2] HALLMARK_TNFA_SIGNALING_VIA_NFKB--HALLMARK_TGF_BETA_SIGNALING</span>
<span class="co">#  [3] HALLMARK_TNFA_SIGNALING_VIA_NFKB--HALLMARK_IL6_JAK_STAT3_SIGNALING</span>
<span class="co">#  [4] HALLMARK_TNFA_SIGNALING_VIA_NFKB--HALLMARK_APOPTOSIS</span>
<span class="co">#  [5] HALLMARK_TNFA_SIGNALING_VIA_NFKB--HALLMARK_MYOGENESIS</span>
<span class="co">#  [6] HALLMARK_TNFA_SIGNALING_VIA_NFKB--HALLMARK_COMPLEMENT</span>
<span class="co">#  [7] HALLMARK_TNFA_SIGNALING_VIA_NFKB--HALLMARK_EPITHELIAL_MESENCHYMAL_TRANSITION</span>
<span class="co">#  [8] HALLMARK_TNFA_SIGNALING_VIA_NFKB--HALLMARK_INFLAMMATORY_RESPONSE</span>
<span class="co"># + ... omitted several edges</span>
<span class="co"># number of geneset clusters:  10</span></code></pre></div>
</div>
<div id="vignette" class="section level2">
<h2 class="hasAnchor">
<a href="#vignette" class="anchor"></a>Vignette</h2>
<p>A vignette is available at <a href="https://clatworthylab.github.io/bowerbird/articles/bowerbird_vignette.html" class="uri">https://clatworthylab.github.io/bowerbird/articles/bowerbird_vignette.html</a></p>
</div>
<div id="examples-of-different-ways-of-loading-in-genesets" class="section level2">
<h2 class="hasAnchor">
<a href="#examples-of-different-ways-of-loading-in-genesets" class="anchor"></a>Examples of different ways of loading in genesets</h2>
<p>Example #1 Doing some manual editing of a predefined geneset.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">bowerbird</span><span class="op">)</span>
<span class="co"># download from msigdb website</span>
<span class="va">file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"c5.go.bp.v7.4.symbols.gmt"</span>, package <span class="op">=</span> <span class="st">"bowerbird"</span><span class="op">)</span>
<span class="co"># manualy read in to do some fine adjustments/filtering</span>
<span class="va">geneset</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/read_geneset.html">read_geneset</a></span><span class="op">(</span><span class="va">file</span><span class="op">)</span> <span class="co"># reads in gene file manually</span>
<span class="co"># do a bit of manual filtering</span>
<span class="va">geneset</span> <span class="op">&lt;-</span> <span class="va">geneset</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html">grep</a></span><span class="op">(</span><span class="st">'B_CELL|T_CELL|NATURAL_KILLER|ANTIBODY|ANTIGEN|LYMPHOCYTE|IMMUNE|INTERFERON|TOLL|INNATE_IMM|ADAPTIVE_IMM'</span>, <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">geneset</span><span class="op">)</span><span class="op">)</span><span class="op">]</span>

<span class="va">bwr</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/bower.html">bower</a></span><span class="op">(</span><span class="va">geneset</span><span class="op">)</span>
<span class="va">bwr</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/snn_graph.html">snn_graph</a></span><span class="op">(</span><span class="va">bwr</span><span class="op">)</span>
<span class="va">bwr</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/find_clusters.html">find_clusters</a></span><span class="op">(</span><span class="va">bwr</span><span class="op">)</span>
<span class="va">bwr</span> <span class="op">&lt;-</span> <span class="fu">summarize_clusters</span><span class="op">(</span><span class="va">bwr</span><span class="op">)</span></code></pre></div>
<p>Example #2 Extracting from <code>msigdbr</code>.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">bowerbird</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://igordot.github.io/msigdbr/">msigdbr</a></span><span class="op">)</span>
<span class="va">GO</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span><span class="fu">msigdbr</span><span class="fu">::</span><span class="fu"><a href="https://igordot.github.io/msigdbr/reference/msigdbr.html">msigdbr</a></span><span class="op">(</span>category <span class="op">=</span> <span class="st">"C5"</span>, subcategory <span class="op">=</span> <span class="st">"GO:BP"</span><span class="op">)</span><span class="op">)</span>
<span class="va">genesets</span> <span class="op">&lt;-</span> <span class="va">GO</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html">grep</a></span><span class="op">(</span><span class="st">'_B_CELL|_T_CELL|NATURAL_KILLER|ANTIBODY|ANTIGEN|LYMPHOCYTE|IMMUNE|INTERFERON|TOLL|INNATE_IMM|ADAPTIVE_IMM'</span>, <span class="va">GO</span><span class="op">$</span><span class="va">gs_name</span><span class="op">)</span>, <span class="op">]</span>

<span class="co"># convert to list</span>
<span class="va">gs_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="va">genesets</span><span class="op">$</span><span class="va">gs_name</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="va">genesets</span><span class="op">[</span><span class="va">genesets</span><span class="op">$</span><span class="va">gs_name</span> <span class="op">%in%</span> <span class="va">x</span>, <span class="st">"gene_symbol"</span><span class="op">]</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">gs_list</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="va">genesets</span><span class="op">$</span><span class="va">gs_name</span><span class="op">)</span>

<span class="va">bwr</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/bower.html">bower</a></span><span class="op">(</span><span class="va">gs_list</span><span class="op">)</span>
<span class="va">bwr</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/snn_graph.html">snn_graph</a></span><span class="op">(</span><span class="va">bwr</span><span class="op">)</span>
<span class="va">bwr</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/find_clusters.html">find_clusters</a></span><span class="op">(</span><span class="va">bwr</span><span class="op">)</span>
<span class="va">bwr</span> <span class="op">&lt;-</span> <span class="fu">summarize_clusters</span><span class="op">(</span><span class="va">bwr</span><span class="op">)</span>

<span class="co"># plot</span>
<span class="fu"><a href="reference/plot_graph.html">plot_graph</a></span><span class="op">(</span><span class="va">bwr</span>, colorby <span class="op">=</span> <span class="st">'cluster'</span>, node.size <span class="op">=</span> <span class="st">'geneset_size'</span><span class="op">)</span></code></pre></div>
</div>
<div id="geneset-enrichment-analysis" class="section level2">
<h2 class="hasAnchor">
<a href="#geneset-enrichment-analysis" class="anchor"></a>geneset enrichment analysis</h2>
<p>To do this, we would need either a single-cell object (<code>SingleCellExperiment</code> or <code>Seurat</code>), or a list containing differential gene testing results</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># load a dummy dataset</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">ktplots</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="va">kidneyimmune</span><span class="op">)</span>

<span class="co"># with single-cell methods:</span>

<span class="co">## AUCell</span>
<span class="va">bwr</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/enrich_genesets.html">enrich_genesets</a></span><span class="op">(</span><span class="va">kidneyimmune</span>, <span class="va">bwr</span>, groupby <span class="op">=</span> <span class="st">'celltype'</span>, mode <span class="op">=</span> <span class="st">'AUCell'</span><span class="op">)</span>

<span class="co">## Seurat::AddModuleScore</span>
<span class="va">bwr</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/enrich_genesets.html">enrich_genesets</a></span><span class="op">(</span><span class="va">kidneyimmune</span>, <span class="va">bwr</span>, groupby <span class="op">=</span> <span class="st">'celltype'</span>, mode <span class="op">=</span> <span class="st">'Seurat'</span><span class="op">)</span>

<span class="co">## scanpy.tl.score_genes</span>
<span class="va">bwr</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/enrich_genesets.html">enrich_genesets</a></span><span class="op">(</span><span class="va">kidneyimmune</span>, <span class="va">bwr</span>, groupby <span class="op">=</span> <span class="st">'celltype'</span>, mode <span class="op">=</span> <span class="st">'scanpy'</span><span class="op">)</span>

<span class="co"># with fgsea on deg tables in a list</span>
<span class="va">degs</span> <span class="op">&lt;-</span> <span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://satijalab.org/seurat/reference/FindAllMarkers.html">FindAllMarkers</a></span><span class="op">(</span><span class="va">kidneyimmune</span><span class="op">)</span>
<span class="va">degs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/split.html">split</a></span><span class="op">(</span><span class="va">degs</span>, <span class="va">degs</span><span class="op">$</span><span class="va">cluster</span><span class="op">)</span> <span class="co"># so in practice, there should be one DEG table per comparison in a list.</span>
<span class="va">bwr</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/enrich_genesets.html">enrich_genesets</a></span><span class="op">(</span><span class="va">degs</span>, <span class="va">bwr</span>, gene_symbol <span class="op">=</span> <span class="st">'gene'</span>, logfoldchanges <span class="op">=</span> <span class="st">'avg_log2FC'</span>,  pvals <span class="op">=</span> <span class="st">'p_val'</span><span class="op">)</span>

<span class="co"># plot</span>
<span class="va">celltypes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/levels.html">levels</a></span><span class="op">(</span><span class="va">kidneyimmune</span><span class="op">)</span> <span class="co"># celtypes stored as Idents in seurat object.</span>
<span class="va">plot_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span><span class="va">celltypes</span>, <span class="kw">function</span><span class="op">(</span><span class="va">ds</span><span class="op">)</span><span class="op">{</span>
  <span class="va">g</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/plot_graph.html">plot_graph</a></span><span class="op">(</span><span class="va">bwr</span>, colorby <span class="op">=</span> <span class="va">ds</span><span class="op">)</span> <span class="op">+</span> <span class="fu">viridis</span><span class="fu">::</span><span class="fu"><a href="https://sjmgarnier.github.io/viridis/reference/scale_viridis.html">scale_color_viridis</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu">theme</span><span class="op">(</span>plot.title <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>size <span class="op">=</span> <span class="fl">8</span>, face <span class="op">=</span> <span class="st">"bold"</span><span class="op">)</span><span class="op">)</span>
<span class="op">}</span><span class="op">)</span>
<span class="fu">cowplot</span><span class="fu">::</span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html">plot_grid</a></span><span class="op">(</span>plotlist <span class="op">=</span> <span class="va">plot_list</span>, scale <span class="op">=</span> <span class="fl">0.9</span><span class="op">)</span></code></pre></div>
</div>
<div id="using-core-genes-for-enrichment" class="section level2">
<h2 class="hasAnchor">
<a href="#using-core-genes-for-enrichment" class="anchor"></a>Using core genes for enrichment</h2>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># same function as above, just specifying core = TRUE.</span>
<span class="va">bwr</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/enrich_genesets.html">enrich_genesets</a></span><span class="op">(</span><span class="va">kidneyimmune</span>, <span class="va">bwr</span>, core <span class="op">=</span> <span class="cn">TRUE</span>, groupby <span class="op">=</span> <span class="st">'celltype'</span><span class="op">)</span> <span class="co"># if mode is not specified, defaults to AUCell.</span>
<span class="co"># output can also be plotted via ggraph as above but i'm using heatmap for convenience.</span>
<span class="fu">pheatmap</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/pheatmap/man/pheatmap.html">pheatmap</a></span><span class="op">(</span><span class="va">bwr</span><span class="op">@</span><span class="va">scores</span>, cellheight <span class="op">=</span> <span class="fl">20</span>, cellwidth <span class="op">=</span> <span class="fl">20</span><span class="op">)</span>

<span class="co"># same for the gsea results.</span>
<span class="va">bwr</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/enrich_genesets.html">enrich_genesets</a></span><span class="op">(</span><span class="va">degs</span>, <span class="va">bwr</span>, core <span class="op">=</span> <span class="cn">TRUE</span>, gene_symbol <span class="op">=</span> <span class="st">'gene'</span>, logfoldchanges <span class="op">=</span> <span class="st">'avg_logFC'</span>,  pvals <span class="op">=</span> <span class="st">'p_val'</span><span class="op">)</span>
<span class="fu">pheatmap</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/pheatmap/man/pheatmap.html">pheatmap</a></span><span class="op">(</span><span class="va">bwr</span><span class="op">@</span><span class="va">scores</span><span class="op">$</span><span class="va">padj</span>, cellheight <span class="op">=</span> <span class="fl">20</span>, cellwidth <span class="op">=</span> <span class="fl">20</span>, color <span class="op">=</span> <span class="fu">viridis</span><span class="fu">::</span><span class="fu"><a href="https://sjmgarnier.github.io/viridisLite/reference/viridis.html">viridis</a></span><span class="op">(</span><span class="fl">50</span>, direction <span class="op">=</span> <span class="op">-</span><span class="fl">1</span><span class="op">)</span><span class="op">)</span></code></pre></div>
</div>
</div>

  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="license">
<h2>License</h2>
<ul class="list-unstyled">
<li>`use_mit_license()`, `use_gpl3_license()` or friends to pick a
    license</li>
</ul>
</div>
<div class="developers">
<h2>Developers</h2>
<ul class="list-unstyled">
<li>Zewen Kelvin Tuong <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0002-6735-6808" target="orcid.widget" aria-label="ORCID"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
</ul>
</div>

  <div class="dev-status">
<h2>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/clatworthylab/bowerbird/actions/workflows/r.yml"><img src="https://github.com/clatworthylab/bowerbird/actions/workflows/r.yml/badge.svg?branch=master" alt="R"></a></li>
<li><a href="https://github.com/clatworthylab/bowerbird/actions/workflows/vignette.yml"><img src="https://github.com/clatworthylab/bowerbird/actions/workflows/vignette.yml/badge.svg" alt="vignette"></a></li>
<li><a href="https://codecov.io/gh/clatworthylab/bowerbird"><img src="https://codecov.io/gh/clatworthylab/bowerbird/branch/master/graph/badge.svg?token=NPXokMhnc2" alt="codecov"></a></li>
</ul>
</div>
</div>
</div>


      <footer><div class="copyright">
  <p>Developed by Zewen Kelvin Tuong.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
