<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Introduction to bowerbird • bowerbird</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Introduction to bowerbird">
<meta property="og:description" content="bowerbird">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">bowerbird</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/bowerbird_vignette.html">Introduction to bowerbird</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="bowerbird_vignette_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Introduction to bowerbird</h1>
            
            <h4 class="date">2021-07-27</h4>
      
      
      <div class="hidden name"><code>bowerbird_vignette.rmd</code></div>

    </div>

    
    
<p><a href="https://github.com/clatworthylab/bowerbird/actions/workflows/r.yml"><img src="https://github.com/clatworthylab/bowerbird/actions/workflows/r.yml/badge.svg?branch=master" alt="R"></a> <a href="https://github.com/clatworthylab/bowerbird/actions/workflows/vignette.yml"><img src="https://github.com/clatworthylab/bowerbird/actions/workflows/vignette.yml/badge.svg" alt="vignette"></a></p>
<div id="bowerbird" class="section level1">
<h1 class="hasAnchor">
<a href="#bowerbird" class="anchor"></a>bowerbird</h1>
<p>A single-cell package for Functional Annotation + Gene Module Summarization.</p>
<p><img src="../img/logo.png" alt="bower" width="200"></p>
<p><em>Bowerbirds are famous for the elaborate and sometimes whimsical structures that males build to court females.</em></p>
<p><a href="https://blog.nature.org/science/2021/01/04/bowerbirds-meet-the-bird-worlds-kleptomaniac-love-architects/"><img src="https://blog.nature.org/science/files/2020/11/32487196918_8dd537c82a_k.jpg" alt="bower" width="400"></a></p>
<p>Picture credits: A satin bowerbird’s bower, decorated with blue objects. © Stefan Marks / <a href="https://www.flickr.com/photos/stefan_marks/32487196918/">Flickr</a></p>
<p>I’m not sure if this mating ritual is in anyway similar to my attempts to visualise pathway analyses and their association with celltype identities/states… who cares. It’s just a name.</p>
<div id="installation" class="section level2">
<h2 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h2>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># because currently private, requires specifying personal github access token</span>
<span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://devtools.r-lib.org//reference/remote-reexports.html">install_github</a></span><span class="op">(</span><span class="st">'clatworthylab/bowerbird'</span>, auth_token <span class="op">=</span> <span class="st">"insert_your_personal_github_access_token"</span><span class="op">)</span>
<span class="co"># or clone this repo and do devtools::install('path/of/folder/to/bowerbird')</span></code></pre></div>
</div>
<div id="quick-usage" class="section level2">
<h2 class="hasAnchor">
<a href="#quick-usage" class="anchor"></a>Quick Usage</h2>
<p>Pathway analyis is tricky; often there’s too many pathways/terms to use and it’s not easy to distinguish how important every element is. <code>bowerbird</code> attempts to summarise your pathways of interest into potentially meaningful functional annotation modules which you can then take forward to pathway analyses, or the reverse situation where you have a set of pathways that were determined <i><b>a priori</b></i> to be significant and you wanted to find out how to interpret it. To do this, we compute a shortest nearest neighbor graph using the genes in the gene sets/pathways and learn a representative summary term from the name of the gene sets that cluster together.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">bowerbird</span><span class="op">)</span>
<span class="va">bwr</span> <span class="op">&lt;-</span> <span class="fu">bower</span><span class="op">(</span><span class="va">gmt_file</span><span class="op">)</span>
<span class="va">bwr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/snn_graph.html">snn_graph</a></span><span class="op">(</span><span class="va">bwr</span><span class="op">)</span>
<span class="va">bwr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/find_clusters.html">find_clusters</a></span><span class="op">(</span><span class="va">bwr</span><span class="op">)</span>
<span class="va">bwr</span> <span class="op">&lt;-</span> <span class="fu">summarize_clusters</span><span class="op">(</span><span class="va">bwr</span><span class="op">)</span>
<span class="va">bwr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/enrich_genesets.html">enrich_genesets</a></span><span class="op">(</span><span class="va">sce</span>, <span class="va">bwr</span><span class="op">)</span></code></pre></div>
</div>
<div id="full-example" class="section level2">
<h2 class="hasAnchor">
<a href="#full-example" class="anchor"></a>Full example</h2>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">bowerbird</span><span class="op">)</span>
<span class="va">gmt_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"h.all.v7.4.symbols.gmt"</span>, package <span class="op">=</span> <span class="st">"bowerbird"</span><span class="op">)</span> <span class="co"># read in a .gmt file</span>
<span class="va">bwr</span> <span class="op">&lt;-</span> <span class="fu">bower</span><span class="op">(</span><span class="va">gmt_file</span><span class="op">)</span> <span class="co"># this performs a read_geneset step internally, which accepts .gmt, .gmx, .csv, .tsv, .txt, or R objects as list or data.frame format.</span>
<span class="va">bwr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/snn_graph.html">snn_graph</a></span><span class="op">(</span><span class="va">bwr</span><span class="op">)</span>
<span class="va">bwr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/find_clusters.html">find_clusters</a></span><span class="op">(</span><span class="va">bwr</span><span class="op">)</span>
<span class="va">bwr</span> <span class="op">&lt;-</span> <span class="fu">summarize_clusters</span><span class="op">(</span><span class="va">bwr</span><span class="op">)</span></code></pre></div>
<p>Input formats accepts .gmt, .gmx, .csv, .tsv, .txt, or R objects as list or data.frame format.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">csv_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"h.all.v7.4.symbols.csv"</span>, package <span class="op">=</span> <span class="st">"bowerbird"</span><span class="op">)</span> <span class="co"># read in a .gmt file</span>
<span class="va">csv</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.csv</a></span><span class="op">(</span><span class="va">csv_file</span><span class="op">)</span>
<span class="co"># see first 6 rows of the first 3 columns</span>
<span class="va">csv</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">6</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">]</span>
<span class="co">#&gt;   HALLMARK_ADIPOGENESIS HALLMARK_ALLOGRAFT_REJECTION HALLMARK_ANDROGEN_RESPONSE</span>
<span class="co">#&gt; 1                 ABCA1                        AARS1                      ABCC4</span>
<span class="co">#&gt; 2                 ABCB8                        ABCE1                      ABHD2</span>
<span class="co">#&gt; 3                 ACAA2                         ABI1                      ACSL3</span>
<span class="co">#&gt; 4                 ACADL                         ACHE                      ACTN1</span>
<span class="co">#&gt; 5                 ACADM                       ACVR2A                    ADAMTS1</span>
<span class="co">#&gt; 6                 ACADS                         AKT1                      ADRM1</span></code></pre></div>
<p>And now to process it through.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">bwr</span> <span class="op">&lt;-</span> <span class="fu">bower</span><span class="op">(</span><span class="va">csv</span><span class="op">)</span>
<span class="va">bwr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/snn_graph.html">snn_graph</a></span><span class="op">(</span><span class="va">bwr</span><span class="op">)</span>
<span class="va">bwr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/find_clusters.html">find_clusters</a></span><span class="op">(</span><span class="va">bwr</span><span class="op">)</span>
<span class="va">bwr</span> <span class="op">&lt;-</span> <span class="fu">summarize_clusters</span><span class="op">(</span><span class="va">bwr</span><span class="op">)</span></code></pre></div>
</div>
<div id="examples-of-different-ways-of-loading-in-genesets" class="section level2">
<h2 class="hasAnchor">
<a href="#examples-of-different-ways-of-loading-in-genesets" class="anchor"></a>Examples of different ways of loading in genesets</h2>
<p>Doing some manual editing of a predefined geneset.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># download from msigdb website</span>
<span class="va">file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"c5.go.bp.v7.4.symbols.gmt"</span>, package <span class="op">=</span> <span class="st">"bowerbird"</span><span class="op">)</span>
<span class="co"># manualy read in to do some fine adjustments/filtering</span>
<span class="va">geneset</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/read_geneset.html">read_geneset</a></span><span class="op">(</span><span class="va">file</span><span class="op">)</span> <span class="co"># reads in gene file manually</span>
<span class="co"># do a bit of manual filtering</span>
<span class="va">geneset</span> <span class="op">&lt;-</span> <span class="va">geneset</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html">grep</a></span><span class="op">(</span><span class="st">'B_CELL|T_CELL|NATURAL_KILLER|ANTIBODY|ANTIGEN|LYMPHOCYTE|IMMUNE|INTERFERON|TOLL|INNATE|ADAPTIVE'</span>, <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">geneset</span><span class="op">)</span><span class="op">)</span><span class="op">]</span>
<span class="va">geneset</span> <span class="op">&lt;-</span> <span class="va">geneset</span><span class="op">[</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html">grepl</a></span><span class="op">(</span><span class="st">'TROPHOBLAST_CELL|FAT_CELL|ENT_CELL'</span>, <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">geneset</span><span class="op">)</span><span class="op">)</span><span class="op">]</span>

<span class="va">bwr</span> <span class="op">&lt;-</span> <span class="fu">bower</span><span class="op">(</span><span class="va">geneset</span><span class="op">)</span>
<span class="va">bwr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/snn_graph.html">snn_graph</a></span><span class="op">(</span><span class="va">bwr</span><span class="op">)</span>
<span class="va">bwr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/find_clusters.html">find_clusters</a></span><span class="op">(</span><span class="va">bwr</span><span class="op">)</span>
<span class="va">bwr</span> <span class="op">&lt;-</span> <span class="fu">summarize_clusters</span><span class="op">(</span><span class="va">bwr</span><span class="op">)</span>
<span class="co">#&gt; Downloading udpipe model from https://raw.githubusercontent.com/jwijffels/udpipe.models.ud.2.5/master/inst/udpipe-ud-2.5-191206/english-ewt-ud-2.5-191206.udpipe to /var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T//Rtmpn5hu8z/english-ewt-ud-2.5-191206.udpipe</span>
<span class="co">#&gt;  - This model has been trained on version 2.5 of data from https://universaldependencies.org</span>
<span class="co">#&gt;  - The model is distributed under the CC-BY-SA-NC license: https://creativecommons.org/licenses/by-nc-sa/4.0</span>
<span class="co">#&gt;  - Visit https://github.com/jwijffels/udpipe.models.ud.2.5 for model license details.</span>
<span class="co">#&gt;  - For a list of all models and their licenses (most models you can download with this package have either a CC-BY-SA or a CC-BY-SA-NC license) read the documentation at ?udpipe_download_model. For building your own models: visit the documentation by typing vignette('udpipe-train', package = 'udpipe')</span>
<span class="co">#&gt; Downloading finished, model stored at '/var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T//Rtmpn5hu8z/english-ewt-ud-2.5-191206.udpipe'</span></code></pre></div>
<p>Or extracting from <code>msigdbr</code>.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://igordot.github.io/msigdbr/">msigdbr</a></span><span class="op">)</span>
<span class="va">GO</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span><span class="fu">msigdbr</span><span class="fu">::</span><span class="fu"><a href="https://igordot.github.io/msigdbr/reference/msigdbr.html">msigdbr</a></span><span class="op">(</span>category <span class="op">=</span> <span class="st">"C5"</span>, subcategory <span class="op">=</span> <span class="st">"GO:BP"</span><span class="op">)</span><span class="op">)</span>
<span class="va">genesets</span> <span class="op">&lt;-</span> <span class="va">GO</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html">grep</a></span><span class="op">(</span><span class="st">'B_CELL|T_CELL|NATURAL_KILLER|ANTIBODY|ANTIGEN|LYMPHOCYTE|IMMUNE|INTERFERON|TOLL|INNATE|ADAPTIVE'</span>, <span class="va">GO</span><span class="op">$</span><span class="va">gs_name</span><span class="op">)</span>, <span class="op">]</span>
<span class="va">genesets</span> <span class="op">&lt;-</span> <span class="va">genesets</span><span class="op">[</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html">grepl</a></span><span class="op">(</span><span class="st">'TROPHOBLAST_CELL|FAT_CELL|ENT_CELL'</span>, <span class="va">genesets</span><span class="op">$</span><span class="va">gs_name</span><span class="op">)</span>, <span class="op">]</span>

<span class="co"># convert to list</span>
<span class="va">gs_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="va">genesets</span><span class="op">$</span><span class="va">gs_name</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="va">genesets</span><span class="op">[</span><span class="va">genesets</span><span class="op">$</span><span class="va">gs_name</span> <span class="op">%in%</span> <span class="va">x</span>, <span class="st">"gene_symbol"</span><span class="op">]</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">gs_list</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="va">genesets</span><span class="op">$</span><span class="va">gs_name</span><span class="op">)</span>

<span class="va">bwr</span> <span class="op">&lt;-</span> <span class="fu">bower</span><span class="op">(</span><span class="va">gs_list</span><span class="op">)</span>
<span class="va">bwr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/snn_graph.html">snn_graph</a></span><span class="op">(</span><span class="va">bwr</span><span class="op">)</span>
<span class="va">bwr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/find_clusters.html">find_clusters</a></span><span class="op">(</span><span class="va">bwr</span><span class="op">)</span>
<span class="va">bwr</span> <span class="op">&lt;-</span> <span class="fu">summarize_clusters</span><span class="op">(</span><span class="va">bwr</span><span class="op">)</span>
<span class="co">#&gt; Downloading udpipe model from https://raw.githubusercontent.com/jwijffels/udpipe.models.ud.2.5/master/inst/udpipe-ud-2.5-191206/english-ewt-ud-2.5-191206.udpipe to /var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T//Rtmpn5hu8z/english-ewt-ud-2.5-191206.udpipe</span>
<span class="co">#&gt;  - This model has been trained on version 2.5 of data from https://universaldependencies.org</span>
<span class="co">#&gt;  - The model is distributed under the CC-BY-SA-NC license: https://creativecommons.org/licenses/by-nc-sa/4.0</span>
<span class="co">#&gt;  - Visit https://github.com/jwijffels/udpipe.models.ud.2.5 for model license details.</span>
<span class="co">#&gt;  - For a list of all models and their licenses (most models you can download with this package have either a CC-BY-SA or a CC-BY-SA-NC license) read the documentation at ?udpipe_download_model. For building your own models: visit the documentation by typing vignette('udpipe-train', package = 'udpipe')</span>
<span class="co">#&gt; Downloading finished, model stored at '/var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T//Rtmpn5hu8z/english-ewt-ud-2.5-191206.udpipe'</span></code></pre></div>
<p>So what has actually happened? Well, basically <code>bowerbird</code> has learnt which gene sets share more genes with each other, clustered them together and learnt a new summary term.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">bwr</span>
<span class="co">#&gt; BOWER class</span>
<span class="co">#&gt; number of genesets:  265 </span>
<span class="co">#&gt; genesets kNN Graph: </span>
<span class="co">#&gt; IGRAPH c77c135 UNW- 265 843 -- </span>
<span class="co">#&gt; + attr: name (v/c), cluster (v/n), geneset_size (v/n), terms (v/c),</span>
<span class="co">#&gt; | labels (v/c), weight (e/n)</span>
<span class="co">#&gt; + edges from c77c135 (vertex names):</span>
<span class="co">#&gt; [1] GOBP_ACTIVATED_T_CELL_PROLIFERATION--GOBP_CD8_POSITIVE_ALPHA_BETA_T_CELL_DIFFERENTIATION          </span>
<span class="co">#&gt; [2] GOBP_ACTIVATED_T_CELL_PROLIFERATION--GOBP_NEGATIVE_REGULATION_OF_ACTIVATED_T_CELL_PROLIFERATION   </span>
<span class="co">#&gt; [3] GOBP_ACTIVATED_T_CELL_PROLIFERATION--GOBP_NEGATIVE_REGULATION_OF_LYMPHOCYTE_DIFFERENTIATION       </span>
<span class="co">#&gt; [4] GOBP_ACTIVATED_T_CELL_PROLIFERATION--GOBP_NEGATIVE_REGULATION_OF_T_CELL_RECEPTOR_SIGNALING_PATHWAY</span>
<span class="co">#&gt; [5] GOBP_ACTIVATION_OF_IMMUNE_RESPONSE --GOBP_ACTIVATION_OF_INNATE_IMMUNE_RESPONSE                    </span>
<span class="co">#&gt; + ... omitted several edges</span>
<span class="co">#&gt; number of geneset clusters:  21 </span>
<span class="co">#&gt; Core genes:</span>
<span class="co">#&gt;  First six genes shown</span>
<span class="co">#&gt;   IMMUNE RESPONSE-1 : AIM2 BCL10 BTRC CARD11 CARD9 CD209 ...</span>
<span class="co">#&gt;   CD4 ALPHA BETA T CELL DIFFERENTIATION : ANXA1 ATP7A BATF BCL3 BCL6 CBFB ...</span>
<span class="co">#&gt;   TYPE 2 IMMUNE RESPONSE : IL4 PLCL2 TLR4 IL18 IL18R1 IL4R ...</span>
<span class="co">#&gt;   IMMUNE RESPONSE : CLNK HLA-E HLA-F IFNA2 IFNB1 IL18 ...</span>
<span class="co">#&gt;   T CELL MEDIATED CYTOTOXICITY : HLA-A HLA-B HLA-C HLA-E HLA-F HLA-G ...</span>
<span class="co">#&gt;   B CELL ACTIVATION : ADA BAX BCL6 BLK BTK CD74 ...</span>
<span class="co">#&gt;   RESPONSE TO INTERFERON BETA : BCL2L11 HIF1A MEF2C ADA SOS1 SOS2 ...</span>
<span class="co">#&gt;   CD8 ALPHA BETA T CELL ACTIVATION : HLA-E GATA3 ADCYAP1 BTK C3 CCR7 ...</span>
<span class="co">#&gt;   T CELL DIFFERENTIATION : FCGR2B FGL2 HLA-DOA LILRB2 THBS1 BMP4 ...</span>
<span class="co">#&gt;   INTERFERON GAMMA PRODUCTION : ABCB9 ACE ACTR10 ACTR1A ACTR1B AP1B1 ...</span>
<span class="co">#&gt;   TOLL LIKE RECEPTOR SIGNALING PATHWAY : CACTIN PTPRS TICAM1 TLR3 TLR4 TLR9 ...</span>
<span class="co">#&gt;   T CELL CYTOKINE PRODUCTION : KLRC1 LILRB1 NOD2 PTPRC EGR3 LEF1 ...</span>
<span class="co">#&gt;   B CELL DIFFERENTIATION : IL10 TLR9 SYK ADA BCL3 C17orf99 ...</span>
<span class="co">#&gt;   ACTIVATED T CELL PROLIFERATION : LILRB4 CD24 CD274 FYN HHLA2 ICOSLG ...</span>
<span class="co">#&gt;   T CELL SELECTION : TOX BCL2 BATF CTSL FOXP3 IL12B ...</span>
<span class="co">#&gt;   LYMPHOCYTE CHEMOTAXIS : CCL2 CCL3 CCL4 CCL5 CCL7 CXCL14 ...</span>
<span class="co">#&gt;   B CELL ACTIVATION-1 : ADA ADAM17 BATF BCL3 BCL6 BTK ...</span>
<span class="co">#&gt;   ALPHA BETA T CELL ACTIVATION : AGER ANXA1 ARG2 ATP7A BATF BCL3 ...</span>
<span class="co">#&gt;   INFLAMMATORY RESPONSE TO ANTIGENIC STIMULUS : BTK C3 CD28 HLA-E LTA NOD2 ...</span>
<span class="co">#&gt;   ANTIGEN PROCESSING AND PRESENTATION OF PEPTIDE ANTIGEN VIA MHC CLASS I : ABCB9 AZGP1 B2M ERAP1 ERAP2 HFE ...</span>
<span class="co">#&gt;   LYMPHOCYTE MIGRATION : ADAM10 ADAM17 ADAM8 AIF1 APP CCL20 ...</span></code></pre></div>
<p>We can visualise it with a simple plot</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plot_graph.html">plot_graph</a></span><span class="op">(</span><span class="va">bwr</span>, colorby <span class="op">=</span> <span class="st">'cluster'</span>, node.size <span class="op">=</span> <span class="st">'geneset_size'</span><span class="op">)</span></code></pre></div>
<p><img src="bowerbird_vignette_files/figure-html/unnamed-chunk-10-1.png" width="672"></p>
</div>
<div id="geneset-average-expression-testing" class="section level2">
<h2 class="hasAnchor">
<a href="#geneset-average-expression-testing" class="anchor"></a>Geneset average expression testing</h2>
<p>We can also compute the enrichment score and superimpose the values onto the visualization. To do this, we would need either a single-cell object (<code>SingleCellExperiment</code> or <code>Seurat</code>), or a list containing differential gene testing results.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># load a dummy dataset</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">ktplots</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="va">kidneyimmune</span><span class="op">)</span></code></pre></div>
<p>With popular single-cell methods.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## AUCell</span>
<span class="va">bwr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/enrich_genesets.html">enrich_genesets</a></span><span class="op">(</span><span class="va">kidneyimmune</span>, <span class="va">bwr</span>, groupby <span class="op">=</span> <span class="st">'celltype'</span>, mode <span class="op">=</span> <span class="st">'AUCell'</span><span class="op">)</span>

<span class="co">## Seurat::AddModuleScore</span>
<span class="va">bwr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/enrich_genesets.html">enrich_genesets</a></span><span class="op">(</span><span class="va">kidneyimmune</span>, <span class="va">bwr</span>, groupby <span class="op">=</span> <span class="st">'celltype'</span>, mode <span class="op">=</span> <span class="st">'Seurat'</span><span class="op">)</span>

<span class="co">## scanpy.tl.score_genes</span>
<span class="va">bwr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/enrich_genesets.html">enrich_genesets</a></span><span class="op">(</span><span class="va">kidneyimmune</span>, <span class="va">bwr</span>, groupby <span class="op">=</span> <span class="st">'celltype'</span>, mode <span class="op">=</span> <span class="st">'scanpy'</span><span class="op">)</span></code></pre></div>
<p>Visualise as a network plot.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">celltypes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/levels.html">levels</a></span><span class="op">(</span><span class="va">kidneyimmune</span><span class="op">)</span> <span class="co"># celtypes stored as Idents in seurat object.</span>
<span class="va">plot_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span><span class="va">celltypes</span>, <span class="kw">function</span><span class="op">(</span><span class="va">ds</span><span class="op">)</span><span class="op">{</span>
  <span class="va">g</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plot_graph.html">plot_graph</a></span><span class="op">(</span><span class="va">bwr</span>, colorby <span class="op">=</span> <span class="va">ds</span><span class="op">)</span> <span class="op">+</span> <span class="fu">viridis</span><span class="fu">::</span><span class="fu"><a href="https://sjmgarnier.github.io/viridis/reference/scale_viridis.html">scale_color_viridis</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu">theme</span><span class="op">(</span>plot.title <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>size <span class="op">=</span> <span class="fl">8</span>, face <span class="op">=</span> <span class="st">"bold"</span><span class="op">)</span><span class="op">)</span>
<span class="op">}</span><span class="op">)</span>
<span class="fu">cowplot</span><span class="fu">::</span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html">plot_grid</a></span><span class="op">(</span>plotlist <span class="op">=</span> <span class="va">plot_list</span>, scale <span class="op">=</span> <span class="fl">0.9</span><span class="op">)</span></code></pre></div>
<p><img src="bowerbird_vignette_files/figure-html/unnamed-chunk-13-1.png" width="4320"></p>
</div>
<div id="running-fgsea-on-deg-tables-in-a-list-as-an-alternative-to-the-single-cell-methods-" class="section level2">
<h2 class="hasAnchor">
<a href="#running-fgsea-on-deg-tables-in-a-list-as-an-alternative-to-the-single-cell-methods-" class="anchor"></a>Running <code>fgsea</code> on deg tables in a list as an alternative to the single-cell methods.</h2>
<p>Here we use a differential expression result for marker genes as input.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">degs</span> <span class="op">&lt;-</span> <span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://satijalab.org/seurat/reference/FindAllMarkers.html">FindAllMarkers</a></span><span class="op">(</span><span class="va">kidneyimmune</span><span class="op">)</span>
<span class="va">degs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/split.html">split</a></span><span class="op">(</span><span class="va">degs</span>, <span class="va">degs</span><span class="op">$</span><span class="va">cluster</span><span class="op">)</span> <span class="co"># so in practice, there should be one DEG table per comparison in a list.</span>
<span class="va">bwr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/enrich_genesets.html">enrich_genesets</a></span><span class="op">(</span><span class="va">degs</span>, <span class="va">bwr</span>, gene_symbol <span class="op">=</span> <span class="st">'gene'</span>, logfoldchanges <span class="op">=</span> <span class="st">'avg_log2FC'</span>,  pvals <span class="op">=</span> <span class="st">'p_val'</span><span class="op">)</span></code></pre></div>
<p>Coloured by adjusted p value.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">plot_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span><span class="va">celltypes</span>, <span class="kw">function</span><span class="op">(</span><span class="va">ds</span><span class="op">)</span><span class="op">{</span>
  <span class="va">g</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plot_graph.html">plot_graph</a></span><span class="op">(</span><span class="va">bwr</span>, colorby <span class="op">=</span> <span class="va">ds</span>, mode <span class="op">=</span> <span class="st">'gsea'</span>, gsea.slot <span class="op">=</span> <span class="st">'padj'</span><span class="op">)</span> <span class="op">+</span> <span class="fu">viridis</span><span class="fu">::</span><span class="fu"><a href="https://sjmgarnier.github.io/viridis/reference/scale_viridis.html">scale_color_viridis</a></span><span class="op">(</span>na.value <span class="op">=</span> <span class="st">'#e7e7e7'</span><span class="op">)</span> <span class="op">+</span> <span class="fu">theme</span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">'none'</span>, plot.title <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>size <span class="op">=</span> <span class="fl">8</span>, face <span class="op">=</span> <span class="st">"bold"</span><span class="op">)</span><span class="op">)</span>
<span class="op">}</span><span class="op">)</span>
<span class="fu">cowplot</span><span class="fu">::</span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html">plot_grid</a></span><span class="op">(</span>plotlist <span class="op">=</span> <span class="va">plot_list</span>, scale <span class="op">=</span> <span class="fl">0.9</span><span class="op">)</span></code></pre></div>
<p><img src="bowerbird_vignette_files/figure-html/unnamed-chunk-15-1.png" width="3840"></p>
<p>Coloured by normalized enrichment score.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">plot_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span><span class="va">celltypes</span>, <span class="kw">function</span><span class="op">(</span><span class="va">ds</span><span class="op">)</span><span class="op">{</span>
  <span class="va">g</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plot_graph.html">plot_graph</a></span><span class="op">(</span><span class="va">bwr</span>, colorby <span class="op">=</span> <span class="va">ds</span>, mode <span class="op">=</span> <span class="st">'gsea'</span>, gsea.slot <span class="op">=</span> <span class="st">'NES'</span><span class="op">)</span> <span class="op">+</span> <span class="fu">viridis</span><span class="fu">::</span><span class="fu"><a href="https://sjmgarnier.github.io/viridis/reference/scale_viridis.html">scale_color_viridis</a></span><span class="op">(</span>na.value <span class="op">=</span> <span class="st">'#e7e7e7'</span><span class="op">)</span> <span class="op">+</span> <span class="fu">theme</span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">'none'</span>, plot.title <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>size <span class="op">=</span> <span class="fl">8</span>, face <span class="op">=</span> <span class="st">"bold"</span><span class="op">)</span><span class="op">)</span>
<span class="op">}</span><span class="op">)</span>
<span class="fu">cowplot</span><span class="fu">::</span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html">plot_grid</a></span><span class="op">(</span>plotlist <span class="op">=</span> <span class="va">plot_list</span>, scale <span class="op">=</span> <span class="fl">0.9</span><span class="op">)</span></code></pre></div>
<p><img src="bowerbird_vignette_files/figure-html/unnamed-chunk-16-1.png" width="3840"></p>
</div>
<div id="using-core-genes-for-enrichment" class="section level2">
<h2 class="hasAnchor">
<a href="#using-core-genes-for-enrichment" class="anchor"></a>Using <code>core</code> genes for enrichment</h2>
<p>Same function as above, just specifying <code>core = TRUE</code>. This will grab the intersecting (or outersecting rather) genes in a gene set cluster and perform the enrichment, like a summary gene set of the summarised terms.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">bwr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/enrich_genesets.html">enrich_genesets</a></span><span class="op">(</span><span class="va">kidneyimmune</span>, <span class="va">bwr</span>, core <span class="op">=</span> <span class="cn">TRUE</span>, groupby <span class="op">=</span> <span class="st">'celltype'</span><span class="op">)</span> <span class="co"># if mode is not specified, the default option is to use AUCell</span>
<span class="fu">pheatmap</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/pheatmap/man/pheatmap.html">pheatmap</a></span><span class="op">(</span><span class="va">bwr</span><span class="op">@</span><span class="va">scores</span>, cellheight <span class="op">=</span> <span class="fl">20</span>, cellwidth <span class="op">=</span> <span class="fl">20</span><span class="op">)</span></code></pre></div>
<p><img src="bowerbird_vignette_files/figure-html/unnamed-chunk-17-1.png" width="1152"></p>
<p>Same for the gsea results. First looking at the NES.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">bwr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/enrich_genesets.html">enrich_genesets</a></span><span class="op">(</span><span class="va">degs</span>, <span class="va">bwr</span>, core <span class="op">=</span> <span class="cn">TRUE</span>, gene_symbol <span class="op">=</span> <span class="st">'gene'</span>, logfoldchanges <span class="op">=</span> <span class="st">'avg_log2FC'</span>,  pvals <span class="op">=</span> <span class="st">'p_val'</span><span class="op">)</span>
<span class="fu">pheatmap</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/pheatmap/man/pheatmap.html">pheatmap</a></span><span class="op">(</span><span class="va">bwr</span><span class="op">@</span><span class="va">scores</span><span class="op">$</span><span class="va">NES</span>, cellheight <span class="op">=</span> <span class="fl">20</span>, cellwidth <span class="op">=</span> <span class="fl">20</span><span class="op">)</span></code></pre></div>
<p><img src="bowerbird_vignette_files/figure-html/unnamed-chunk-18-1.png" width="1152"></p>
<p>And now the FDR, remembering that <code>&lt;0.25</code> is usually the cut off.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">pheatmap</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/pheatmap/man/pheatmap.html">pheatmap</a></span><span class="op">(</span><span class="va">bwr</span><span class="op">@</span><span class="va">scores</span><span class="op">$</span><span class="va">padj</span>, cellheight <span class="op">=</span> <span class="fl">20</span>, cellwidth <span class="op">=</span> <span class="fl">20</span>, color <span class="op">=</span> <span class="fu">viridis</span><span class="fu">::</span><span class="fu"><a href="https://sjmgarnier.github.io/viridisLite/reference/viridis.html">viridis</a></span><span class="op">(</span><span class="fl">50</span>, direction <span class="op">=</span> <span class="op">-</span><span class="fl">1</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><img src="bowerbird_vignette_files/figure-html/unnamed-chunk-19-1.png" width="1152"></p>
<p>To visualise as a graph, can consider doing:</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">bwr2</span> <span class="op">&lt;-</span> <span class="fu">bower</span><span class="op">(</span><span class="va">bwr</span><span class="op">@</span><span class="va">coregenes</span><span class="op">)</span>
<span class="co"># and so on</span></code></pre></div>
<p>Hopefully you find this interesting and potentially useful.</p>
<p>Please email me at <a href="mailto:kt16@sanger.ac.uk" class="email">kt16@sanger.ac.uk</a> if you have any queries, feedback or ideas and I’ll get back to you!</p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Zewen Kelvin Tuong.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
